File: Week-4/Day-2/Problem4.py
# A transformation sequence from word beginWord to word endWord using a dictionary wordList is a sequence of words beginWord -> s1 -> s2 -> ... -> sk such that:

# Every adjacent pair of words differs by a single letter.
# Every si for 1 <= i <= k is in wordList. Note that beginWord does not need to be in wordList.
# sk == endWord
# Given two words, beginWord and endWord, and a dictionary wordList, return all the shortest transformation sequences from beginWord to endWord, or an empty list if no such sequence exists. Each sequence should be returned as a list of the words [beginWord, s1, s2, ..., sk].

# Example 1:

# Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]
# Output: [["hit","hot","dot","dog","cog"],["hit","hot","lot","log","cog"]]
# Explanation: There are 2 shortest transformation sequences:
# "hit" -> "hot" -> "dot" -> "dog" -> "cog"
# "hit" -> "hot" -> "lot" -> "log" -> "cog"
# Example 2:

# Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log"]
# Output: []
# Explanation: The endWord "cog" is not in wordList, therefore there is no valid transformation sequence.
 
# Constraints:

# 1 <= beginWord.length <= 5
# endWord.length == beginWord.length
# 1 <= wordList.length <= 500
# wordList[i].length == beginWord.length
# beginWord, endWord, and wordList[i] consist of lowercase English letters.
# beginWord != endWord
# All the words in wordList are unique.
# The sum of all shortest transformation sequences does not exceed 105.


# https://leetcode.com/problems/word-ladder-ii/


class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        wordSet = set(wordList)
        if endWord not in wordSet:
            return []
        
        graph = collections.defaultdict(list)
        queue = collections.deque([beginWord])
        visited = set([beginWord])
        found = False
        
        while queue and not found:
            currentLevelVisited = set()
            for _ in range(len(queue)):
                word = queue.popleft()
                for i in range(len(word)):
                    for c in 'abcdefghijklmnopqrstuvwxyz':
                        newWord = word[:i] + c + word[i + 1:]
                        if newWord in wordSet:
                            if newWord not in visited:
                                currentLevelVisited.add(newWord)
                                queue.append(newWord)
                            graph[word].append(newWord)
                            if newWord == endWord:
                                found = True
            
            visited.update(currentLevelVisited)

        if not found:
            return []

        result = []

        def backtrack(currentWord, path):
            if currentWord == beginWord:
                result.append(path[::-1])
                return
            for neighbor in graph[currentWord]:
                backtrack(neighbor, path + [neighbor])

        backtrack(endWord, [endWord])
        return result

File: Week-7/Day-5/problem-4.swift
class Solution {
    func findLadders(_ beginWord: String, _ endWord: String, _ wordList: [String]) -> [[String]] {
        var wordSet = Set(wordList)
        var results = [[String]]()
        var queue = [(beginWord, [beginWord])]
        var visited = Set([beginWord])
        var found = false
        
        while !queue.isEmpty && !found {
            var currentLevelVisited = Set<String>()
            var currentLevelQueue = [(String, [String])]()
            
            for (word, path) in queue {
                if word == endWord {
                    results.append(path)
                    found = true
                }
                for i in 0..<word.count {
                    let wordArray = Array(word)
                    for char in "abcdefghijklmnopqrstuvwxyz" {
                        var newWord = wordArray
                        newWord[i] = char
                        let newWordStr = String(newWord)
                        if wordSet.contains(newWordStr) && !visited.contains(newWordStr) {
                            currentLevelQueue.append((newWordStr, path + [newWordStr]))
                            currentLevelVisited.insert(newWordStr)
                        }
                    }
                }
            }
            
            visited.formUnion(currentLevelVisited)
            queue = currentLevelQueue
        }
        
        return results
    }
}
